File: tools/FabricDataFactoryMigrationAssistant/src/services/activityTransformer.ts
import { connectionService } from './connectionService';
import { copyActivityTransformer } from './copyActivityTransformer';
import { customActivityTransformer } from './customActivityTransformer';

export class ActivityTransformer {
  // Transform LinkedService references to Fabric format - comprehensive implementation
  transformLinkedServiceReferencesToFabric(activity: any, pipelineConnectionMappings?: any): void {
    if (!activity || typeof activity !== 'object') return;

    // Skip Copy activities - they are handled by specialized copyActivityTransformer
    if (activity.type === 'Copy') {
      return; // Do not process Copy activities here
    }

    // Skip Custom activities - they are handled by specialized customActivityTransformer
    if (activity.type === 'Custom') {
      return; // Do not process Custom activities here
    }

    // 1. Remove ADF-specific linkedServiceName references and replace with Fabric externalReferences
    this.removeLinkedServiceReferencesAndSetExternalReferences(activity);

    // 2. Convert static text properties to Expression objects for Fabric compatibility
    this.convertStaticTextToExpressions(activity);

    // 3. Add required properties for different activity types
    this.addRequiredPropertiesForActivityType(activity);

    // 4. Handle dataset references that contain LinkedService references
    this.transformDatasetReferences(activity);
  }

  removeLinkedServiceReferencesAndSetExternalReferences(activity: any): void {
    if (!activity) return;

    let connectionId: string | undefined;
    let linkedServiceName: string | undefined;

    const typeProperties = activity.typeProperties || {};

    if (typeProperties.linkedServices && Array.isArray(typeProperties.linkedServices)) {
      const firstLinkedService = typeProperties.linkedServices[0];
      if (firstLinkedService?.referenceName) {
        linkedServiceName = firstLinkedService.referenceName;
        connectionId = connectionService.mapLinkedServiceToConnection(linkedServiceName);
        delete typeProperties.linkedServices;
      }
    }

    if (typeProperties.linkedServiceName?.referenceName) {
      linkedServiceName = typeProperties.linkedServiceName.referenceName;
      connectionId = connectionService.mapLinkedServiceToConnection(linkedServiceName);
      delete typeProperties.linkedServiceName;
    }

    if (activity.linkedServiceName?.referenceName) {
      linkedServiceName = activity.linkedServiceName.referenceName;
      connectionId = connectionService.mapLinkedServiceToConnection(linkedServiceName);
      delete activity.linkedServiceName;
    }

    if (connectionId) {
      if (!activity.externalReferences) activity.externalReferences = {};
      activity.externalReferences.connection = connectionId;
    } else if (linkedServiceName) {
      console.warn(`No connection mapping found for LinkedService: ${linkedServiceName} in activity ${activity.name}`);
    }
  }

  convertStaticTextToExpressions(activity: any): void {
    if (!activity?.typeProperties || typeof activity.typeProperties !== 'object') return;

    const typeProperties = activity.typeProperties;

    switch (activity.type) {
      case 'Script': this.convertScriptActivityExpressions(typeProperties); break;
      case 'StoredProcedure': this.convertStoredProcedureActivityExpressions(typeProperties); break;
      case 'WebActivity': this.convertWebActivityExpressions(typeProperties); break;
      case 'Lookup': this.convertLookupActivityExpressions(typeProperties); break;
      default: this.convertCommonStringPropertiesToExpressions(typeProperties); break;
    }
  }

  convertScriptActivityExpressions(typeProperties: any): void {
    if (typeProperties.scripts && Array.isArray(typeProperties.scripts)) {
      typeProperties.scripts = typeProperties.scripts.map((script: any) => {
        if (script.text && typeof script.text === 'string') {
          return { ...script, text: { value: script.text, type: 'Expression' } };
        }
        return script;
      });
    }
  }

  convertStoredProcedureActivityExpressions(typeProperties: any): void {
    if (typeProperties.storedProcedureName && typeof typeProperties.storedProcedureName === 'string') {
      typeProperties.storedProcedureName = { value: typeProperties.storedProcedureName, type: 'Expression' };
    }

    if (typeProperties.storedProcedureParameters && typeof typeProperties.storedProcedureParameters === 'object') {
      for (const [paramName, paramValue] of Object.entries(typeProperties.storedProcedureParameters)) {
        if (typeof paramValue === 'string') {
          typeProperties.storedProcedureParameters[paramName] = { value: paramValue, type: 'Expression' };
        }
      }
    }
  }

  convertWebActivityExpressions(typeProperties: any): void {
    if (typeProperties.url && typeof typeProperties.url === 'string') {
      typeProperties.url = { value: typeProperties.url, type: 'Expression' };
    }
    if (typeProperties.body && typeof typeProperties.body === 'string') {
      typeProperties.body = { value: typeProperties.body, type: 'Expression' };
    }
  }

  convertLookupActivityExpressions(typeProperties: any): void {
    if (typeProperties.source?.query && typeof typeProperties.source.query === 'string') {
      typeProperties.source.query = { value: typeProperties.source.query, type: 'Expression' };
    }
  }

  convertCommonStringPropertiesToExpressions(typeProperties: any): void {
    const commonStringProperties = ['query', 'command', 'script', 'sql', 'statement'];
    for (const prop of commonStringProperties) {
      if (typeProperties[prop] && typeof typeProperties[prop] === 'string') {
        typeProperties[prop] = { value: typeProperties[prop], type: 'Expression' };
      }
    }
  }

  addRequiredPropertiesForActivityType(activity: any): void {
    if (!activity?.typeProperties || typeof activity.typeProperties !== 'object') return;

    switch (activity.type) {
      case 'Script': this.addScriptActivityRequiredProperties(activity); break;
      case 'StoredProcedure': this.addStoredProcedureActivityRequiredProperties(activity); break;
      case 'WebActivity': this.addWebActivityRequiredProperties(activity); break;
      default: this.addCommonRequiredProperties(activity); break;
    }
  }

  addScriptActivityRequiredProperties(activity: any): void {
    const typeProperties = activity.typeProperties;
    
    // Only add scriptBlockExecutionTimeout if not present
    if (!typeProperties.scriptBlockExecutionTimeout) {
      typeProperties.scriptBlockExecutionTimeout = '02:00:00';
    }

    // DO NOT add default database - extract from original LinkedService/Dataset if needed
    // but never add placeholder values
    if (!typeProperties.database) {
      // Try to extract database from the original activity's linked service or dataset
      const databaseName = this.extractDatabaseFromActivityReferences(activity);
      if (databaseName) {
        typeProperties.database = databaseName;
        console.log(`Extracted database name '${databaseName}' for Script activity '${activity.name}'`);
      }
      // If no database is found, DON'T add a default - let Fabric handle it
    }
  }

  addStoredProcedureActivityRequiredProperties(activity: any): void {
    const typeProperties = activity.typeProperties;
    
    // DO NOT add default database - extract from original LinkedService/Dataset if needed
    // but never add placeholder values
    if (!typeProperties.database) {
      // Try to extract database from the original activity's linked service or dataset
      const databaseName = this.extractDatabaseFromActivityReferences(activity);
      if (databaseName) {
        typeProperties.database = databaseName;
        console.log(`Extracted database name '${databaseName}' for StoredProcedure activity '${activity.name}'`);
      }
      // If no database is found, DON'T add a default - let Fabric handle it
    }
  }

  /**
   * Extracts database name from activity's LinkedService or Dataset references
   * @param activity The activity to extract database from
   * @returns The database name if found, undefined otherwise
   */
  extractDatabaseFromActivityReferences(activity: any): string | undefined {
    try {
      // Import the adfParserService to access parsed components
      const { adfParserService } = require('./adfParserService');
      
      // For now, we'll implement a basic version that doesn't add defaults
      // This method can be enhanced to extract from actual LinkedService definitions
      // stored in the ADF parser service
      
      // Check if there's a database property in the activity's typeProperties already
      if (activity.typeProperties?.database) {
        return activity.typeProperties.database;
      }
      
      // Try to get from dataset if this is a Copy activity with dataset settings
      if (activity.typeProperties?.source?.datasetSettings?.typeProperties?.database) {
        return activity.typeProperties.source.datasetSettings.typeProperties.database;
      }
      
      if (activity.typeProperties?.sink?.datasetSettings?.typeProperties?.database) {
        return activity.typeProperties.sink.datasetSettings.typeProperties.database;
      }
      
      // For activities with direct LinkedService references, try to extract database
      const linkedServiceRef = this.getLinkedServiceReference(activity);
      if (linkedServiceRef && adfParserService) {
        const linkedService = adfParserService.getLinkedServiceByName(linkedServiceRef);
        if (linkedService?.definition?.properties?.typeProperties) {
          const linkedServiceProps = linkedService.definition.properties.typeProperties;
          
          // Different linked service types may store database info in different properties
          if (linkedServiceProps.database) {
            return linkedServiceProps.database;
          }
          
          // For SQL Server linked services, check connectionString
          if (linkedServiceProps.connectionString || linkedServiceProps.server) {
            // Extract database from connection string if present
            const database = this.extractDatabaseFromConnectionString(linkedServiceProps.connectionString);
            if (database) {
              return database;
            }
          }
        }
      }
      
      // For other activities, try to extract from dataset references
      if (activity.inputs && Array.isArray(activity.inputs) && activity.inputs.length > 0) {
        const firstInput = activity.inputs[0];
        if (firstInput?.referenceName && adfParserService) {
          const dataset = adfParserService.getDatasetByName(firstInput.referenceName);
          if (dataset?.definition?.properties?.typeProperties) {
            const datasetProps = dataset.definition.properties.typeProperties;
            if (datasetProps.database) {
              return datasetProps.database;
            }
            // For SQL datasets, try other common property names
            if (datasetProps.schema && datasetProps.table) {
              // In some cases, database might be implied by the connection context
              // For now, return undefined to avoid adding defaults
            }
          }
        }
      }
      
      // For now, return undefined to avoid adding default values
      return undefined;
    } catch (error) {
      console.warn('Error extracting database from activity references:', error);
      return undefined;
    }
  }

  /**
   * Extracts the LinkedService reference name from an activity
   * @param activity The activity to extract from
   * @returns The LinkedService reference name if found
   */
  private getLinkedServiceReference(activity: any): string | undefined {
    const typeProperties = activity.typeProperties || {};
    
    // Check various possible locations for LinkedService references
    if (typeProperties.linkedServiceName?.referenceName) {
      return typeProperties.linkedServiceName.referenceName;
    }
    
    if (typeProperties.linkedServices && Array.isArray(typeProperties.linkedServices) && typeProperties.linkedServices.length > 0) {
      return typeProperties.linkedServices[0]?.referenceName;
    }
    
    if (activity.linkedServiceName?.referenceName) {
      return activity.linkedServiceName.referenceName;
    }
    
    return undefined;
  }

  /**
   * Extracts database name from a connection string
   * @param connectionString The connection string to parse
   * @returns The database name if found
   */
  private extractDatabaseFromConnectionString(connectionString: string | any): string | undefined {
    if (!connectionString) return undefined;
    
    // Handle connection string that might be an object with value property
    let connStr: string;
    if (typeof connectionString === 'object' && connectionString.value) {
      connStr = connectionString.value;
    } else if (typeof connectionString === 'string') {
      connStr = connectionString;
    } else {
      return undefined;
    }
    
    // Common patterns for database in connection strings
    const patterns = [
      /database=([^;]+)/i,
      /initial catalog=([^;]+)/i,
      /catalog=([^;]+)/i
    ];
    
    for (const pattern of patterns) {
      const match = connStr.match(pattern);
      if (match && match[1]) {
        return match[1].trim();
      }
    }
    
    return undefined;
  }

  addWebActivityRequiredProperties(activity: any): void {
    const typeProperties = activity.typeProperties;
    if (!typeProperties.method) typeProperties.method = 'GET';
    if (!typeProperties.headers) typeProperties.headers = {};
  }

  addCommonRequiredProperties(activity: any): void {
    if (!activity.policy) activity.policy = {};
    if (!activity.policy.timeout) activity.policy.timeout = '0.12:00:00';
  }

  transformDatasetReferences(activity: any): void {
    if (activity.inputs && Array.isArray(activity.inputs)) activity.inputs = this.transformActivityInputs(activity.inputs);
    if (activity.outputs && Array.isArray(activity.outputs)) activity.outputs = this.transformActivityOutputs(activity.outputs);
    if (activity.type === 'Copy' && activity.typeProperties) this.transformCopyActivityDatasetReferences(activity.typeProperties);
  }

  transformCopyActivityDatasetReferences(typeProperties: any): void {
    if (typeProperties.source) typeProperties.source = this.transformDatasetToDatasetSettings(typeProperties.source, 'source');
    if (typeProperties.sink) typeProperties.sink = this.transformDatasetToDatasetSettings(typeProperties.sink, 'sink');
  }

  transformDatasetToDatasetSettings(datasetConfig: any, role: 'source' | 'sink'): any {
    if (!datasetConfig || typeof datasetConfig !== 'object') return datasetConfig;
    if (datasetConfig.dataset?.referenceName) {
      const datasetName = datasetConfig.dataset.referenceName;
      const connectionId = this.getConnectionIdForDataset(datasetName);
      return { ...datasetConfig, datasetSettings: { ...datasetConfig, externalReferences: connectionId ? { connection: connectionId } : undefined } };
    }
    return datasetConfig;
  }

  // Normalize and transform activity inputs (datasets) into Fabric-friendly datasetSettings
  transformActivityInputs(inputs: any[]): any[] {
    if (!Array.isArray(inputs)) return inputs;

    return inputs.map(input => {
      if (!input) return input;

      // A common ADF format: { "referenceName": "name", "type": "DatasetReference" }
      if (input.type === 'DatasetReference' && (input.referenceName || input.dataset?.referenceName)) {
        const name = input.referenceName || input.dataset?.referenceName;
        const datasetConfig = input.dataset ? input : { dataset: { referenceName: name } };
        const transformed = this.transformDatasetToDatasetSettings(datasetConfig, 'source');
        return { ...transformed, type: 'DatasetReference', referenceName: name };
      }

      // Some inputs already have a dataset sub-object
      if (input.dataset?.referenceName) {
        return this.transformDatasetToDatasetSettings(input, 'source');
      }

      return input;
    });
  }

  // Normalize and transform activity outputs (datasets) into Fabric-friendly datasetSettings
  transformActivityOutputs(outputs: any[]): any[] {
    if (!Array.isArray(outputs)) return outputs;

    return outputs.map(output => {
      if (!output) return output;

      if (output.type === 'DatasetReference' && (output.referenceName || output.dataset?.referenceName)) {
        const name = output.referenceName || output.dataset?.referenceName;
        const datasetConfig = output.dataset ? output : { dataset: { referenceName: name } };
        const transformed = this.transformDatasetToDatasetSettings(datasetConfig, 'sink');
        return { ...transformed, type: 'DatasetReference', referenceName: name };
      }

      if (output.dataset?.referenceName) {
        return this.transformDatasetToDatasetSettings(output, 'sink');
      }

      return output;
    });
  }

  getDatabaseNameFromConnection(connectionId: string): string | undefined {
    console.log(`Getting database name for connection: ${connectionId}`);
    return undefined;
  }

  getConnectionIdForDataset(datasetName: string): string | undefined {
    console.log(`Getting connection ID for dataset: ${datasetName}`);
    return undefined;
  }

  setExternalReferences(activity: any): void {
    if (!activity) return;
    if (!activity.externalReferences) activity.externalReferences = {};
    let connectionId: string | undefined;

    if (activity.typeProperties) {
      if (activity.typeProperties.linkedServices && Array.isArray(activity.typeProperties.linkedServices)) {
        const firstLinkedService = activity.typeProperties.linkedServices[0];
        if (firstLinkedService?.referenceName) {
          connectionId = connectionService.mapLinkedServiceToConnection(firstLinkedService.referenceName);
          delete activity.typeProperties.linkedServices;
        }
      }

      if (activity.typeProperties.linkedServiceName?.referenceName) {
        connectionId = connectionService.mapLinkedServiceToConnection(activity.typeProperties.linkedServiceName.referenceName);
        delete activity.typeProperties.linkedServiceName;
      }

      if (activity.typeProperties.source?.linkedServiceName?.referenceName) {
        connectionId = connectionService.mapLinkedServiceToConnection(activity.typeProperties.source.linkedServiceName.referenceName);
      }

      if (activity.typeProperties.sink?.linkedServiceName?.referenceName) {
        connectionId = connectionService.mapLinkedServiceToConnection(activity.typeProperties.sink.linkedServiceName.referenceName);
      }
    }

    if (connectionId) {
      activity.externalReferences.connection = connectionId;
      console.log(`Set externalReferences.connection to ${connectionId} for activity ${activity.name}`);
    } else {
      console.warn(`No connection mapping found for activity ${activity.name}`);
    }
  }

  updateTypePropertiesLinkedServiceReferences(typeProperties: any): void {
    if (!typeProperties || typeof typeProperties !== 'object') return;

    if (typeProperties.source?.linkedServiceName?.referenceName) {
      const fabricConnectionId = connectionService.mapLinkedServiceToConnection(typeProperties.source.linkedServiceName.referenceName);
      if (fabricConnectionId) typeProperties.source.fabricConnectionId = fabricConnectionId;
    }

    if (typeProperties.sink?.linkedServiceName?.referenceName) {
      const fabricConnectionId = connectionService.mapLinkedServiceToConnection(typeProperties.sink.linkedServiceName.referenceName);
      if (fabricConnectionId) typeProperties.sink.fabricConnectionId = fabricConnectionId;
    }

    if (typeProperties.linkedServiceName?.referenceName) {
      const fabricConnectionId = connectionService.mapLinkedServiceToConnection(typeProperties.linkedServiceName.referenceName);
      if (fabricConnectionId) typeProperties.fabricConnectionId = fabricConnectionId;
    }

    if (typeProperties.dataset?.linkedServiceName?.referenceName) {
      const fabricConnectionId = connectionService.mapLinkedServiceToConnection(typeProperties.dataset.linkedServiceName.referenceName);
      if (fabricConnectionId) typeProperties.dataset.fabricConnectionId = fabricConnectionId;
    }
  }

  activityReferencesFailedConnector(activity: any): boolean {
    if (!activity || typeof activity !== 'object') return false;

    // Special handling for Custom activities
    if (activity.type === 'Custom') {
      return customActivityTransformer.activityReferencesFailedConnector(activity);
    }

    const typeProperties = activity.typeProperties || {};
    const datasets = [typeProperties.dataset, typeProperties.source?.dataset, typeProperties.sink?.dataset, ...(typeProperties.datasets || [])].filter(Boolean);
    for (const dataset of datasets) {
      if (dataset?.linkedServiceName?.referenceName) {
        const linkedServiceName = dataset.linkedServiceName.referenceName;
        if (connectionService.getFailedConnectors().has(linkedServiceName)) return true;
      }
    }

    const linkedServices = [typeProperties.linkedServiceName, typeProperties.linkedService, ...(typeProperties.linkedServices || [])].filter(Boolean);
    for (const linkedService of linkedServices) {
      const referenceName = linkedService.referenceName || linkedService;
      if (typeof referenceName === 'string' && connectionService.getFailedConnectors().has(referenceName)) return true;
    }

    if (typeProperties.source?.linkedServiceName?.referenceName) {
      if (connectionService.getFailedConnectors().has(typeProperties.source.linkedServiceName.referenceName)) return true;
    }
    if (typeProperties.sink?.linkedServiceName?.referenceName) {
      if (connectionService.getFailedConnectors().has(typeProperties.sink.linkedServiceName.referenceName)) return true;
    }

    return false;
  }

  countInactiveActivities(activities: any[]): number {
    return activities.filter(activity => activity.state === 'Inactive').length;
  }

  hasLinkedServiceReferences(activity: any): boolean {
    if (!activity || typeof activity !== 'object') return false;
    const hasInputDatasets = Array.isArray(activity.inputs) && activity.inputs.some((input: any) => input?.type === 'DatasetReference');
    const hasOutputDatasets = Array.isArray(activity.outputs) && activity.outputs.some((output: any) => output?.type === 'DatasetReference');
    const typeProperties = activity.typeProperties || {};
    const hasDirectLinkedServiceRefs = this.hasDirectLinkedServiceReferences(typeProperties);
    return hasInputDatasets || hasOutputDatasets || hasDirectLinkedServiceRefs;
  }

  hasDirectLinkedServiceReferences(typeProperties: any): boolean {
    if (!typeProperties || typeof typeProperties !== 'object') return false;
    const checks = [
      typeProperties.linkedServiceName,
      typeProperties.linkedService,
      typeProperties.source?.linkedServiceName,
      typeProperties.sink?.linkedServiceName,
      typeProperties.dataset?.linkedServiceName,
      ...(typeProperties.linkedServices || [])
    ];
    return checks.some(ref => ref && (ref.referenceName || typeof ref === 'string'));
  }
}

export const activityTransformer = new ActivityTransformer();

File: tools/FabricDataFactoryMigrationAssistant/adf_fabric_migrator/activity_transformer.py
"""
Activity Transformer for ADF to Fabric migration.

This module provides transformation logic for individual ADF activities
to Microsoft Fabric format.
"""

import copy
import json
import logging
from typing import Any, Dict, List, Optional, Set

logger = logging.getLogger(__name__)


class ActivityTransformer:
    """
    Transformer for ADF activities to Fabric format.
    
    This class handles the transformation of individual activities,
    including LinkedService reference conversion, expression transformation,
    and activity-specific property adjustments.
    """
    
    def __init__(self, enable_databricks_to_trident: bool = False):
        """
        Initialize the activity transformer.
        
        Args:
            enable_databricks_to_trident: When True, transforms DatabricksNotebook 
                activities to TridentNotebook format with appropriate property mappings.
        """
        self._failed_connectors: Set[str] = set()
        self._connection_mappings: Dict[str, str] = {}
        self._enable_databricks_to_trident = enable_databricks_to_trident
    
    def set_databricks_to_trident(self, enabled: bool) -> None:
        """
        Enable or disable DatabricksNotebook to TridentNotebook transformation.
        
        Args:
            enabled: When True, transforms DatabricksNotebook activities to TridentNotebook.
        """
        self._enable_databricks_to_trident = enabled
    
    def set_connection_mappings(self, mappings: Dict[str, str]) -> None:
        """
        Set connection mappings for LinkedService to Fabric connection conversion.
        
        Args:
            mappings: Dictionary mapping LinkedService names to Fabric connection IDs.
        """
        self._connection_mappings = mappings
    
    def set_failed_connectors(self, failed: Set[str]) -> None:
        """
        Set the list of failed connectors.
        
        Args:
            failed: Set of LinkedService names that failed to deploy.
        """
        self._failed_connectors = failed
    
    def get_failed_connectors(self) -> Set[str]:
        """Get the set of failed connectors."""
        return self._failed_connectors
    
    def map_linked_service_to_connection(self, linked_service_name: str) -> Optional[str]:
        """
        Map a LinkedService name to a Fabric connection ID.
        
        Args:
            linked_service_name: The ADF LinkedService name.
            
        Returns:
            The Fabric connection ID, or None if not found.
        """
        return self._connection_mappings.get(linked_service_name)
    
    def transform_activity(
        self, 
        activity: Dict[str, Any],
        pipeline_name: str = "unknown"
    ) -> Dict[str, Any]:
        """
        Transform a single ADF activity to Fabric format.
        
        Args:
            activity: The ADF activity definition.
            pipeline_name: Name of the parent pipeline (for context).
            
        Returns:
            Transformed activity definition.
        """
        if not activity or not isinstance(activity, dict):
            return activity
        
        activity_type = activity.get("type", "")
        
        # Skip Copy and Custom activities - they have specialized transformers
        if activity_type in ("Copy", "Custom"):
            return activity
        
        # Transform DatabricksNotebook to TridentNotebook if enabled
        if self._enable_databricks_to_trident and activity_type == "DatabricksNotebook":
            activity = self._transform_databricks_to_trident(activity)
        
        # Transform LinkedService references
        self._remove_linked_service_references_and_set_external_refs(activity)
        
        # Convert static text to expressions for Fabric compatibility
        self._convert_static_text_to_expressions(activity)
        
        # Add required properties for activity type
        self._add_required_properties(activity)
        
        # Transform dataset references
        self._transform_dataset_references(activity)
        
        return activity
    
    def _remove_linked_service_references_and_set_external_refs(
        self, 
        activity: Dict[str, Any]
    ) -> None:
        """Remove ADF LinkedService references and set Fabric externalReferences."""
        connection_id: Optional[str] = None
        linked_service_name: Optional[str] = None
        
        type_properties = activity.get("typeProperties", {})
        
        # Check various locations for LinkedService references
        if isinstance(type_properties.get("linkedServices"), list):
            linked_services = type_properties["linkedServices"]
            if linked_services and linked_services[0].get("referenceName"):
                linked_service_name = linked_services[0]["referenceName"]
                connection_id = self.map_linked_service_to_connection(linked_service_name)
                del type_properties["linkedServices"]
        
        if type_properties.get("linkedServiceName", {}).get("referenceName"):
            linked_service_name = type_properties["linkedServiceName"]["referenceName"]
            connection_id = self.map_linked_service_to_connection(linked_service_name)
            del type_properties["linkedServiceName"]
        
        if activity.get("linkedServiceName", {}).get("referenceName"):
            linked_service_name = activity["linkedServiceName"]["referenceName"]
            connection_id = self.map_linked_service_to_connection(linked_service_name)
            del activity["linkedServiceName"]
        
        if connection_id:
            if "externalReferences" not in activity:
                activity["externalReferences"] = {}
            activity["externalReferences"]["connection"] = connection_id
        elif linked_service_name:
            logger.warning(
                f"No connection mapping found for LinkedService: {linked_service_name} "
                f"in activity {activity.get('name')}"
            )
    
    def _convert_static_text_to_expressions(self, activity: Dict[str, Any]) -> None:
        """Convert static text properties to Expression objects for Fabric."""
        type_properties = activity.get("typeProperties", {})
        if not isinstance(type_properties, dict):
            return
        
        activity_type = activity.get("type", "")
        
        if activity_type == "Script":
            self._convert_script_expressions(type_properties)
        elif activity_type == "StoredProcedure":
            self._convert_stored_procedure_expressions(type_properties)
        elif activity_type == "WebActivity":
            self._convert_web_activity_expressions(type_properties)
        elif activity_type == "Lookup":
            self._convert_lookup_expressions(type_properties)
        else:
            self._convert_common_string_expressions(type_properties)
    
    def _transform_databricks_to_trident(self, activity: Dict[str, Any]) -> Dict[str, Any]:
        """
        Transform DatabricksNotebook activity to TridentNotebook for Fabric.
        
        This transformation:
        - Changes type from "DatabricksNotebook" to "TridentNotebook"
        - Removes: linkedServiceName, notebookPath
        - Adds: notebookId and workspaceId inside typeProperties (placeholders)
        - Renames "baseParameters" to "parameters"
        - Uses double nesting structure for parameters:
          "param": { "value": { "value": "@expr", "type": "Expression" }, "type": "Expression" }
        
        Args:
            activity: The DatabricksNotebook activity definition.
            
        Returns:
            Transformed TridentNotebook activity definition.
        """
        transformed = copy.deepcopy(activity)
        
        # Change type
        transformed["type"] = "TridentNotebook"
        
        # Get typeProperties
        type_properties = transformed.get("typeProperties", {})
        
        # Remove notebookPath
        if "notebookPath" in type_properties:
            del type_properties["notebookPath"]
        
        # Remove linkedServiceName from activity and typeProperties
        if "linkedServiceName" in transformed:
            del transformed["linkedServiceName"]
        if "linkedServiceName" in type_properties:
            del type_properties["linkedServiceName"]
        
        # Add placeholders for notebookId and workspaceId
        type_properties["notebookId"] = "<PLACEHOLDER_NOTEBOOK_ID>"
        type_properties["workspaceId"] = "<PLACEHOLDER_WORKSPACE_ID>"
        
        # Transform baseParameters to parameters with double nesting
        if "baseParameters" in type_properties:
            base_params = type_properties["baseParameters"]
            new_params = {}
            
            if isinstance(base_params, dict):
                for param_name, param_value in base_params.items():
                    # Apply double nesting structure
                    new_params[param_name] = self._create_double_nested_expression(param_value)
            
            type_properties["parameters"] = new_params
            del type_properties["baseParameters"]
        
        transformed["typeProperties"] = type_properties
        
        logger.info(
            f"Transformed DatabricksNotebook '{activity.get('name')}' to TridentNotebook"
        )
        
        return transformed
    
    def _create_double_nested_expression(self, value: Any) -> Dict[str, Any]:
        """
        Create double nested expression structure for TridentNotebook parameters.
        
        Creates structure: { "value": { "value": "@expr", "type": "Expression" }, "type": "Expression" }
        
        Args:
            value: The original parameter value.
            
        Returns:
            Double nested expression structure.
        """
        # Determine the inner value
        if isinstance(value, dict):
            # If it's already an expression object, extract the value
            if value.get("type") == "Expression" and "value" in value:
                inner_value = value["value"]
            else:
                # Convert dict to string representation
                inner_value = json.dumps(value)
        else:
            inner_value = str(value) if value is not None else ""
        
        # Create double nested structure
        return {
            "value": {
                "value": inner_value,
                "type": "Expression"
            },
            "type": "Expression"
        }
    
    def _convert_script_expressions(self, type_properties: Dict[str, Any]) -> None:
        """Convert Script activity expressions."""
        scripts = type_properties.get("scripts", [])
        if isinstance(scripts, list):
            type_properties["scripts"] = [
                {**script, "text": {"value": script["text"], "type": "Expression"}}
                if isinstance(script.get("text"), str)
                else script
                for script in scripts
            ]
    
    def _convert_stored_procedure_expressions(self, type_properties: Dict[str, Any]) -> None:
        """Convert StoredProcedure activity expressions."""
        if isinstance(type_properties.get("storedProcedureName"), str):
            type_properties["storedProcedureName"] = {
                "value": type_properties["storedProcedureName"],
                "type": "Expression",
            }
        
        params = type_properties.get("storedProcedureParameters", {})
        if isinstance(params, dict):
            for param_name, param_value in params.items():
                if isinstance(param_value, str):
                    params[param_name] = {"value": param_value, "type": "Expression"}
    
    def _convert_web_activity_expressions(self, type_properties: Dict[str, Any]) -> None:
        """Convert WebActivity expressions."""
        if isinstance(type_properties.get("url"), str):
            type_properties["url"] = {"value": type_properties["url"], "type": "Expression"}
        
        if isinstance(type_properties.get("body"), str):
            type_properties["body"] = {"value": type_properties["body"], "type": "Expression"}
    
    def _convert_lookup_expressions(self, type_properties: Dict[str, Any]) -> None:
        """Convert Lookup activity expressions."""
        source = type_properties.get("source", {})
        if isinstance(source, dict) and isinstance(source.get("query"), str):
            source["query"] = {"value": source["query"], "type": "Expression"}
    
    def _convert_common_string_expressions(self, type_properties: Dict[str, Any]) -> None:
        """Convert common string properties to expressions."""
        common_props = ["query", "command", "script", "sql", "statement"]
        for prop in common_props:
            if isinstance(type_properties.get(prop), str):
                type_properties[prop] = {
                    "value": type_properties[prop], 
                    "type": "Expression"
                }
    
    def _add_required_properties(self, activity: Dict[str, Any]) -> None:
        """Add required properties for activity type."""
        type_properties = activity.get("typeProperties", {})
        if not isinstance(type_properties, dict):
            return
        
        activity_type = activity.get("type", "")
        
        if activity_type == "Script":
            if "scriptBlockExecutionTimeout" not in type_properties:
                type_properties["scriptBlockExecutionTimeout"] = "02:00:00"
        elif activity_type == "WebActivity":
            if "method" not in type_properties:
                type_properties["method"] = "GET"
            if "headers" not in type_properties:
                type_properties["headers"] = {}
        
        # Add common required properties
        if "policy" not in activity:
            activity["policy"] = {}
        if "timeout" not in activity["policy"]:
            activity["policy"]["timeout"] = "0.12:00:00"
    
    def _transform_dataset_references(self, activity: Dict[str, Any]) -> None:
        """Transform dataset references in activity."""
        if "inputs" in activity and isinstance(activity["inputs"], list):
            activity["inputs"] = self.transform_activity_inputs(activity["inputs"])
        
        if "outputs" in activity and isinstance(activity["outputs"], list):
            activity["outputs"] = self.transform_activity_outputs(activity["outputs"])
    
    def transform_activity_inputs(self, inputs: List[Any]) -> List[Any]:
        """Transform activity inputs (datasets)."""
        if not isinstance(inputs, list):
            return inputs
        
        return [
            self._transform_dataset_reference(inp, "source")
            for inp in inputs
        ]
    
    def transform_activity_outputs(self, outputs: List[Any]) -> List[Any]:
        """Transform activity outputs (datasets)."""
        if not isinstance(outputs, list):
            return outputs
        
        return [
            self._transform_dataset_reference(out, "sink")
            for out in outputs
        ]
    
    def _transform_dataset_reference(
        self, 
        dataset_ref: Any, 
        role: str
    ) -> Any:
        """Transform a single dataset reference."""
        if not isinstance(dataset_ref, dict):
            return dataset_ref
        
        # Handle DatasetReference type
        if dataset_ref.get("type") == "DatasetReference":
            ref_name = dataset_ref.get("referenceName") or dataset_ref.get("dataset", {}).get("referenceName")
            if ref_name:
                return {
                    **dataset_ref,
                    "type": "DatasetReference",
                    "referenceName": ref_name,
                }
        
        return dataset_ref
    
    def activity_references_failed_connector(self, activity: Dict[str, Any]) -> bool:
        """
        Check if an activity references a failed connector.
        
        Args:
            activity: The activity definition.
            
        Returns:
            True if the activity references a failed connector.
        """
        if not activity or not isinstance(activity, dict):
            return False
        
        type_properties = activity.get("typeProperties", {})
        
        # Check dataset references
        datasets = [
            type_properties.get("dataset"),
            type_properties.get("source", {}).get("dataset") if isinstance(type_properties.get("source"), dict) else None,
            type_properties.get("sink", {}).get("dataset") if isinstance(type_properties.get("sink"), dict) else None,
        ]
        datasets.extend(type_properties.get("datasets", []) or [])
        
        for dataset in datasets:
            if isinstance(dataset, dict):
                ls_name = dataset.get("linkedServiceName", {}).get("referenceName")
                if ls_name and ls_name in self._failed_connectors:
                    return True
        
        # Check LinkedService references
        linked_services = [
            type_properties.get("linkedServiceName"),
            type_properties.get("linkedService"),
        ]
        linked_services.extend(type_properties.get("linkedServices", []) or [])
        
        for ls in linked_services:
            if isinstance(ls, dict):
                ref_name = ls.get("referenceName")
            elif isinstance(ls, str):
                ref_name = ls
            else:
                continue
            
            if ref_name and ref_name in self._failed_connectors:
                return True
        
        return False
    
    def count_inactive_activities(self, activities: List[Any]) -> int:
        """Count the number of inactive activities."""
        return sum(
            1 for a in activities 
            if isinstance(a, dict) and a.get("state") == "Inactive"
        )
    
    def has_linked_service_references(self, activity: Dict[str, Any]) -> bool:
        """Check if an activity has LinkedService references."""
        if not isinstance(activity, dict):
            return False
        
        # Check inputs/outputs
        has_input_datasets = any(
            isinstance(inp, dict) and inp.get("type") == "DatasetReference"
            for inp in activity.get("inputs", [])
        )
        has_output_datasets = any(
            isinstance(out, dict) and out.get("type") == "DatasetReference"
            for out in activity.get("outputs", [])
        )
        
        # Check direct references
        type_properties = activity.get("typeProperties", {})
        has_direct_refs = any([
            type_properties.get("linkedServiceName"),
            type_properties.get("linkedService"),
            type_properties.get("source", {}).get("linkedServiceName") if isinstance(type_properties.get("source"), dict) else None,
            type_properties.get("sink", {}).get("linkedServiceName") if isinstance(type_properties.get("sink"), dict) else None,
        ])
        
        return has_input_datasets or has_output_datasets or has_direct_refs


# Singleton instance for convenience
activity_transformer = ActivityTransformer()
